File: DL-5TM/DL-5TM.php
<?php
/* https://www.decentlab.com/products/legacy-soil-moisture-and-temperature-sensor-for-lorawan */

abstract class DecentlabDecoder
{
    const PROTOCOL_VERSION = 2;

    protected $sensors;

    public function decode($payload = '')
    {
        $bytes = hex2bin($payload);
        $parts = [];

        $parts['version'] = ord($bytes[0]);
        if ($parts['version'] != self::PROTOCOL_VERSION) {
            $parts['error'] = sprintf("protocol version %u doesn't match v2", $parts['version']);
            return $parts;
        }

        $parts['device_id'] = unpack('n', substr($bytes, 1))[1];
        $flags = unpack('n', substr($bytes, 3))[1];

        /* decode payload */
        $k = 5;
        foreach ($this->sensors as $sensor) {
            if (($flags & 1) == 1) {
                $x = [];
                /* convert data to 16-bit integer array */
                for ($j = 0; $j < $sensor['length']; $j++) {
                    array_push($x, unpack('n', substr($bytes, $k))[1]);
                    $k += 2;
                }

                /* decode sensor values */
                foreach ($sensor['values'] as $value) {
                    if ($value['convert'] != NULL) {
                        $parts[$value['name'] . '_value'] = $value['convert']($x);
                        if ($value['unit'] != NULL) {
                            $parts[$value['name'] . '_unit'] = $value['unit'];
                        }
                    }
                }
            }
            $flags >>= 1;
        }

        return $parts;
    }
}

class DL_5TM_Decoder extends DecentlabDecoder {
    
    public function __construct()
    {
        $this->sensors = [
            [
                'length' => 2,
                'values' => [
                    [
                        'name' => 'dielectric_permittivity',
                        'convert' => function ($x) { return $x[0] / 50; },
                        'unit' => NULL,
                    ],
                    [
                        'name' => 'volumetric_water_content',
                        'convert' => function ($x) { return 0.0000043 * pow($x[0]/50, 3) - 0.00055 * pow($x[0]/50, 2) + 0.0292 * ($x[0]/50) - 0.053; },
                        'unit' => 'm³⋅m⁻³',
                    ],
                    [
                        'name' => 'soil_temperature',
                        'convert' => function ($x) { return ($x[1] - 400) / 10; },
                        'unit' => '°C',
                    ],
                ],
            ],
            [
                'length' => 1,
                'values' => [
                    [
                        'name' => 'battery_voltage',
                        'convert' => function ($x) { return $x[0] / 1000; },
                        'unit' => 'V',
                    ],
                ],
            ],
        ];
    }
}


$decoder = new DL_5TM_Decoder();
$payloads = [
    '02023b0003003702710c60',
    '02023b00020c60',
];

foreach($payloads as $payload) {
    var_dump($decoder->decode($payload));
}

File: decoders/DL-5TM/DL-5TM.ex

# https://www.decentlab.com/products/legacy-soil-moisture-and-temperature-sensor-for-lorawan

defmodule DecentlabDecoder do
  @protocol_version 2

  defp sensor_defs do
    [
      %{
        length: 2,
        values: [
          %{
            :name => "Dielectric permittivity",
            :convert => fn x -> Enum.at(x, 0) / 50 end,
            :unit => nil
          },
          %{
            :name => "Volumetric water content",
            :convert => fn x -> 0.0000043 * :math.pow(Enum.at(x, 0)/50, 3) - 0.00055 * :math.pow(Enum.at(x, 0)/50, 2) + 0.0292 * (Enum.at(x, 0)/50) - 0.053 end,
            :unit => "m³⋅m⁻³"
          },
          %{
            :name => "Soil temperature",
            :convert => fn x -> (Enum.at(x, 1) - 400) / 10 end,
            :unit => "°C"
          }
        ]
      },
      %{
        length: 1,
        values: [
          %{
            :name => "Battery voltage",
            :convert => fn x -> Enum.at(x, 0) / 1000 end,
            :unit => "V"
          }
        ]
      }
    ]
  end


  def decode(msg, :hex) do
    {:ok, bytes} = Base.decode16(msg, case: :mixed)
    decode(bytes)
  end

  def decode(msg) when is_binary(msg), do: decode_binary(msg)

  def decode(msg), do: to_string(msg) |> decode

  defp decode_binary(<<@protocol_version, device_id::size(16), flags::binary-size(2), bytes::binary>>) do
    bytes
    |> bytes_to_words()
    |> sensor(flags, sensor_defs())
    |> Map.put("Device ID", device_id)
    |> Map.put("Protocol version", @protocol_version)
  end


  defp bytes_to_words(<<>>), do: []

  defp bytes_to_words(<<word::size(16), rest::binary>>), do: [word | bytes_to_words(rest)]


  defp sensor(words, <<flags::size(15), 1::size(1)>>, [%{length: len, values: value_defs} | rest]) do
    {x, rest_words} = Enum.split(words, len)
    value(value_defs, x)
    |> Map.merge(sensor(rest_words, <<0::size(1), flags::size(15)>>, rest))
  end

  defp sensor(words, <<flags::size(15), 0::size(1)>>, [_cur | rest]) do
    sensor(words, <<0::size(1), flags::size(15)>>, rest)
  end

  defp sensor([], _flags, []), do: %{}


  defp value([], _x), do: %{}

  defp value([%{convert: nil} | rest], x), do: value(rest, x)

  defp value([%{name: name, unit: unit, convert: convert} | rest], x) do
    value(rest, x)
    |> Map.put(name, %{"unit" => unit, "value" => convert.(x)})
  end

  defp where(true, if_true, _if_false), do: if_true
  defp where(false, _if_true, if_false), do: if_false

end

IO.inspect(DecentlabDecoder.decode("02023b0003003702710c60", :hex))
IO.inspect(DecentlabDecoder.decode("02023b00020c60", :hex))
